gms/gossiper.cc

```diff
@@ -802,6 +757,84 @@ future<gossiper::endpoint_permit> gossiper::lock_endpoint(inet_address ep) {
     });
 }
 
+
+future<> gossiper::failure_detector_loop_for_node(gms::inet_address node) {
+    auto last = gossiper::clk::now();
+    auto max_duration = std::chrono::milliseconds(_cfg.failure_detector_timeout_in_ms());
+    auto diff = std::chrono::milliseconds(0);
+    auto nr = _live_endpoints.size();
+    int64_t generation = endpoint_state_map[get_broadcast_address()].get_heart_beat_state().get_generation();
+    while (is_enabled()) {
+        bool failed = false;
+        try {
+            logger.debug("failure_detector_loop: Send echo to node {}", node);
+            co_await _messaging.send_gossip_echo(netw::msg_addr(node), generation, max_duration);
+            logger.debug("failure_detector_loop: Send echo to node {}, status = ok", node);
+        } catch (...) {
+            failed = true;
+            logger.info("failure_detector_loop: Send echo to node {}, status = failed: {}", node, std::current_exception());
+        }
+        auto now = gossiper::clk::now();
+        diff = now - last;
+        if (!failed) {
+            last = now;
+        }
+        if (diff > max_duration) {
```

Comment:
If you don't check `failed`, then you can consider a node as up even if your calls keep failing (not due to timeout).

Or if they timeout a bit ahead of `gossiper::clk`.

Reply:
But eventually we will fail, in the next iteration, so there is no problem.